[CLOB-930] Run three other instances of the application during e2e tests that check for non-determinism

lcwik opened the pull request:
### Changelist
Add 3 additional validator invariant checkers (PR) for almost all existing e2e tests:
* parallelApp is responsible for seeing all CheckTx requests, block proposals, blocks, and RecheckTx requests. This allows it to detect state differences due to inconsistent in-memory structures (for example iteration order in maps).
* noCheckTxApp is responsible for seeing all block proposals and blocks. This allows it to simulate a validator that never received any of the CheckTx requests and that it will still accept blocks and arrive at the same state hash.
* crashingApp is responsible for restarting before processing a block and sees all CheckTx requests, block proposals, and blocks. This allows it to check that in memory state can be restored successfully on application restart and that it will accept a block after a crash and arrive at the same state hash.

### Test Plan
Existing tests pass beyond the few places where I had to disable non-determinism checks in certain tests that mutate keeper state directly in a way where non-determinism would be falsely detected or they rely on in-memory state that can't survive app restart.

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

teddyding commented on protocol/app/app.go:
```diff
@@ -212,6 +213,8 @@ type App struct {
 	txConfig          client.TxConfig
 	interfaceRegistry types.InterfaceRegistry
+	db                dbm.DB
+	snapshotDB        dbm.DB
```
For my understanding, where/how is this `snapshotDB` used?

teddyding commented on protocol/testing/e2e/funding/funding_e2e_test.go:
```diff
@@ -141,11 +141,15 @@ func TestFunding(t *testing.T) {
+			tApp := testapp.NewTestAppBuilder(t).
+				// UpdateIndexPrice only contacts the tApp.App.Server causing non-determinism in the
+				// other App instances in TestApp used for non-determinism checking.
+				WithNonDeterminismChecksEnabled(false).
```
For e2e tests that require setting up in-memory state, instead of disabling non-determinism checks altogether (zero coverage), would it make sense to apply the required in-memory setup to the 3 new instances (at least some coverage)? 

For instance, allow test cases to pass in `func SetupApp(app *app.App)` so changes like `pricefeed_testutil.UpdateIndexPrice(` can be applied to not just `tApp.App` but the 3 other instances. 

In the case of the funding e2e test, the better way is probably to overwrite `Prepare/ProcessProposal` to accept the new oracle price, but unsure if other tests could benefit from above.

lcwik replied:
For things that mutate keepers, we should rely on changing the tests to use genesis/CheckTx instead.

For queries, we could expose new capabilities like how TestApp.CheckTx is invoked, we could have something like TestApp.ServerMsg that invokes it on all the instances but I'm pretty undecided as to what to do here as this change moves the needle pretty far.

teddyding replied:
> For things that mutate keepers, we should rely on changing the tests to use genesis/CheckTx instead.

For tests where oracle prices need to be be updated more than once during the lifespan, `genesis` isn't sufficient. I guess in this case we should use `DeliverTxsOverride` to force a `MsgUpdatePrice` transaction?

lcwik replied:
We could ensure that they source the same pricing information.

teddyding commented on protocol/testutil/app/app.go:
```diff
@@ -273,18 +276,44 @@ type ExecuteCheckTxs func(ctx sdk.Context, app *app.App) (stop bool)
+//   - `parallelApp` is responsible for seeing all CheckTx requests, block proposals, blocks, and RecheckTx requests.
+//     This allows it to detect state differences due to inconsistent in-memory structures (for example iteration order
+//     in maps).
+//   - `noCheckTxApp` is responsible for seeing all block proposals and blocks. This allows it to simulate a validator
```
In similar spirit, would it make sense to have a `noisyCheckTxApp` that just takes arbitrary `checkTx` that aren't sent to the main `tApp.App`? This is getting closer to fuzz testing but could be valuable.

lcwik replied:
Yeah, we could totally do that. I don't think we are limited to 3 invariant checkers. Feel free to file a ticket with additional invariants that you can think of.

teddyding replied:
Filed https://linear.app/dydx/issue/CLOB-998/create-a-noisychecktx-app-instance-in-e2e-test-suite

teddyding commented:
https://github.com/dydxprotocol/v4-chain/actions/runs/6671376530/job/18134582211?pr=704 simapp failing and I can reproduce locally. Seems related to DB closing?

lcwik commented on protocol/app/app.go:
```diff
@@ -212,6 +213,8 @@ type App struct {
 	txConfig          client.TxConfig
 	interfaceRegistry types.InterfaceRegistry
+	db                dbm.DB
+	snapshotDB        dbm.DB
```
app.Close() is the location where we should be shutting down resources on application shutdown.

The DB backs the snapshot store and Cosmos SDK doesn't do this for us so we need to shut it down.

lcwik commented on protocol/testutil/app/app.go:
```diff
@@ -382,49 +455,181 @@ func (tApp *TestApp) initChainIfNeeded() {
-	}
-	tApp.App = app
+	// Launch the main instance of the application
+	{
```
The issue that I didn't find a good way to handle the `restart fn` since I rely on updating the references that are captured outside which are shared between the `restart fn` and `cleanup fn`.

I filed https://linear.app/dydx/issue/CORE-721

lcwik commented:
> https://github.com/dydxprotocol/v4-chain/actions/runs/6671376530/job/18134582211?pr=704 simapp failing and I can reproduce locally. Seems related to DB closing?

Turns out that `NewSimApp` is calling App.Close() which is shutting the db down and the call of sim app is also adding a `db.Close()` which is causing a double close. I removed the second close and guarded the Close() call so it can only have effect once. I also updated the test to use clean-up instead of defer to ensure that the Close() call can happen to shutdown the DB before we try to delete everything.
